File: Rime皮肤编辑器/local/local_server.py
#!/usr/bin/env python3
import argparse
import json
import mimetypes
import os
import secrets
import subprocess
import sys
import urllib.parse
import webbrowser
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

BACKUP_ROOT = 'Rime皮肤编辑器备份'
FRONTEND_FILES = {'squirrel.custom.yaml', 'weasel.custom.yaml', 'default.custom.yaml'}
STATIC_ROOT = Path(__file__).resolve().parents[1]


def resolve_allowed_path(root, requested_path):
    raw = str(requested_path or '')
    if not raw or raw.startswith('/') or raw.startswith('\\') or '\0' in raw:
        raise ValueError('不允许访问这个路径。')
    parts = [part for part in raw.replace('\\', '/').split('/') if part]
    if '..' in parts:
        raise ValueError('不允许访问这个路径。')

    is_frontend_file = len(parts) == 1 and parts[0] in FRONTEND_FILES
    is_schema_file = len(parts) == 1 and parts[0].endswith('.schema.yaml')
    is_backup_file = len(parts) >= 2 and parts[0] == BACKUP_ROOT
    if not is_frontend_file and not is_schema_file and not is_backup_file:
        raise ValueError('不允许访问这个路径。')

    root_path = Path(root).resolve()
    target = (root_path / Path(*parts)).resolve()
    try:
        target.relative_to(root_path)
    except ValueError as exc:
        raise ValueError('不允许访问这个路径。') from exc
    return target


def read_config_snapshot(root):
    root_path = Path(root).resolve()
    raw_files = {}
    file_exists = {}
    for key, filename in {
        'squirrel': 'squirrel.custom.yaml',
        'weasel': 'weasel.custom.yaml',
        'default': 'default.custom.yaml',
    }.items():
        file_path = resolve_allowed_path(root_path, filename)
        exists = file_path.exists()
        file_exists[key] = exists
        raw_files[key] = file_path.read_text(encoding='utf-8') if exists else ''
    return {
        'folderName': root_path.name or str(root_path),
        'rawFiles': raw_files,
        'fileExists': file_exists,
        'hasAnySchemaFile': any(path.is_file() and path.name.endswith('.schema.yaml') for path in root_path.iterdir()),
    }


def list_backups(root):
    root_path = Path(root).resolve()
    backup_root = root_path / BACKUP_ROOT
    if not backup_root.exists():
        return []
    backups = []
    for entry in backup_root.iterdir():
        if not entry.is_dir():
            continue
        files = sorted(path.name for path in entry.iterdir() if path.is_file())
        manifest_path = entry / 'manifest.json'
        manifest = read_manifest(manifest_path)
        backups.append({'name': entry.name, 'manifest': manifest, 'availableFiles': files})
    backups.sort(key=lambda item: item['name'], reverse=True)
    return backups


def read_manifest(manifest_path):
    if not manifest_path.exists():
        return None
    try:
        return json.loads(manifest_path.read_text(encoding='utf-8'))
    except json.JSONDecodeError:
        return None


class RimeEditorHandler(BaseHTTPRequestHandler):
    root = None
    token = ''
    static_root = STATIC_ROOT

    def do_GET(self):
        self.handle_request()

    def do_PUT(self):
        self.handle_request()

    def do_POST(self):
        self.handle_request()

    def do_DELETE(self):
        self.handle_request()

    def log_message(self, format, *args):
        return

    def handle_request(self):
        try:
            parsed = urllib.parse.urlparse(self.path)
            if parsed.path.startswith('/api/'):
                if self.headers.get('x-rime-editor-token') != self.token:
                    self.send_json(403, {'error': '访问令牌无效。'})
                    return
                self.handle_api(parsed)
                return
            self.serve_static(parsed.path)
        except Exception as exc:
            self.send_json(400, {'error': str(exc) or '请求失败。'})

    def handle_api(self, parsed):
        query = urllib.parse.parse_qs(parsed.query)
        if self.command == 'GET' and parsed.path == '/api/config':
            self.send_json(200, read_config_snapshot(self.root))
            return
        if self.command == 'GET' and parsed.path == '/api/backups':
            self.send_json(200, {'backups': list_backups(self.root)})
            return
        if self.command == 'GET' and parsed.path == '/api/file':
            path = query.get('path', [''])[0]
            target = resolve_allowed_path(self.root, path)
            if not target.exists():
                self.send_json(404, {'error': '文件不存在。'})
                return
            self.send_json(200, {'exists': True, 'text': target.read_text(encoding='utf-8')})
            return
        if self.command == 'PUT' and parsed.path == '/api/file':
            body = self.read_json_body()
            target = resolve_allowed_path(self.root, body.get('path', ''))
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_text(str(body.get('text', '')), encoding='utf-8')
            self.send_json(200, {'ok': True})
            return
        if self.command == 'DELETE' and parsed.path == '/api/file':
            path = query.get('path', [''])[0]
            target = resolve_allowed_path(self.root, path)
            if target.exists():
                target.unlink()
            self.send_json(200, {'ok': True})
            return
        if self.command == 'POST' and parsed.path == '/api/mkdir':
            body = self.read_json_body()
            target = resolve_allowed_path(self.root, f"{body.get('path', '')}/manifest.json")
            target.parent.mkdir(parents=True, exist_ok=True)
            self.send_json(200, {'ok': True})
            return
        self.send_json(404, {'error': '接口不存在。'})

    def serve_static(self, pathname):
        requested = urllib.parse.unquote(pathname)
        if requested == '/':
            requested = '/index.html'
        relative = requested.lstrip('/')
        target = (self.static_root / relative).resolve()
        try:
            target.relative_to(self.static_root)
        except ValueError:
            self.send_response(404)
            self.end_headers()
            return
        if not target.is_file():
            self.send_response(404)
            self.end_headers()
            return
        content_type = mimetypes.guess_type(str(target))[0] or 'application/octet-stream'
        if content_type.startswith('text/') or target.suffix in {'.js', '.json'}:
            content_type += '; charset=utf-8'
        self.send_response(200)
        self.send_header('content-type', content_type)
        self.send_header('cache-control', 'no-store')
        self.end_headers()
        with target.open('rb') as file:
            self.wfile.write(file.read())

    def read_json_body(self):
        length = int(self.headers.get('content-length') or 0)
        if length > 2 * 1024 * 1024:
            raise ValueError('请求内容过大。')
        if not length:
            return {}
        return json.loads(self.rfile.read(length).decode('utf-8'))

    def send_json(self, status, value):
        body = json.dumps(value, ensure_ascii=False).encode('utf-8')
        self.send_response(status)
        self.send_header('content-type', 'application/json; charset=utf-8')
        self.send_header('cache-control', 'no-store')
        self.send_header('content-length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)


def open_url(url):
    try:
        if sys.platform == 'darwin':
            subprocess.Popen(['open', url])
        elif os.name == 'nt':
            os.startfile(url)  # noqa: S606
        else:
            webbrowser.open(url)
    except Exception:
        webbrowser.open(url)


def main():
    parser = argparse.ArgumentParser(description='Rime 皮肤编辑器本地启动器')
    parser.add_argument('--root', default=str(STATIC_ROOT.parent))
    parser.add_argument('--port', type=int, default=0)
    parser.add_argument('--no-open', action='store_true')
    args = parser.parse_args()

    root = Path(args.root).resolve()
    token = secrets.token_hex(18)
    handler = type('ConfiguredRimeEditorHandler', (RimeEditorHandler,), {
        'root': root,
        'token': token,
        'static_root': STATIC_ROOT,
    })
    server = ThreadingHTTPServer(('127.0.0.1', args.port), handler)
    port = server.server_address[1]
    url = f'http://127.0.0.1:{port}/?token={token}'
    print(f'Rime 皮肤编辑器已启动：{url}', flush=True)
    print(f'配置目录：{root}', flush=True)
    print('关闭这个窗口即可停止本地服务。', flush=True)
    if not args.no_open:
        open_url(url)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('\n已停止本地服务。', flush=True)
    finally:
        server.server_close()


if __name__ == '__main__':
    main()

File: Rime皮肤编辑器/local/local_server.mjs
import {
  createReadStream,
  existsSync,
  lstatSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  realpathSync,
  rmSync,
  statSync,
  writeFileSync,
} from 'node:fs';
import { createServer } from 'node:http';
import { basename, dirname, extname, isAbsolute, join, normalize, relative, resolve, sep } from 'node:path';
import { fileURLToPath } from 'node:url';
import { randomBytes } from 'node:crypto';
import { spawn } from 'node:child_process';

const BACKUP_ROOT = 'Rime皮肤编辑器备份';
const FRONTEND_FILES = new Set(['squirrel.custom.yaml', 'weasel.custom.yaml', 'default.custom.yaml']);
const STATIC_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const MODULE_PATH = fileURLToPath(import.meta.url);
const MIME_TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
};

export function resolveAllowedPath(root, requestedPath) {
  const rootPath = realpathSync(resolve(root));
  const raw = String(requestedPath || '');
  if (!raw || raw.includes('\0') || raw.startsWith('/') || raw.startsWith('\\')) {
    throw new Error('不允许访问这个路径。');
  }

  const rawParts = raw.split(/[\\/]+/);
  if (rawParts.includes('..')) {
    throw new Error('不允许访问这个路径。');
  }

  const normalized = normalize(raw);
  if (normalized === '..' || normalized.startsWith(`..${sep}`) || normalized.includes(`${sep}..${sep}`)) {
    throw new Error('不允许访问这个路径。');
  }

  const parts = normalized.split(/[\\/]+/);
  const isFrontendFile = parts.length === 1 && FRONTEND_FILES.has(parts[0]);
  const isSchemaFile = parts.length === 1 && parts[0].endsWith('.schema.yaml');
  const isBackupFile = parts[0] === BACKUP_ROOT && parts.length >= 2;
  if (!isFrontendFile && !isSchemaFile && !isBackupFile) {
    throw new Error('不允许访问这个路径。');
  }

  const target = resolve(rootPath, normalized);
  const relation = relative(rootPath, target);
  if (relation === '..' || relation.startsWith(`..${sep}`) || resolve(relation) === relation) {
    throw new Error('不允许访问这个路径。');
  }
  if (isSymbolicLink(target)) {
    throw new Error('不允许访问这个路径。');
  }
  const realParent = existingParentRealPath(dirname(target));
  if (!realParent || !isInsideRoot(rootPath, realParent)) {
    throw new Error('不允许访问这个路径。');
  }
  return target;
}

function isSymbolicLink(path) {
  try {
    return lstatSync(path).isSymbolicLink();
  } catch {
    return false;
  }
}

function existingParentRealPath(path) {
  let current = path;
  while (!existsSync(current)) {
    const parent = dirname(current);
    if (parent === current) return null;
    current = parent;
  }
  return realpathSync(current);
}

function isInsideRoot(rootPath, targetPath) {
  const relation = relative(rootPath, targetPath);
  return relation === '' || (!relation.startsWith(`..${sep}`) && relation !== '..' && !isAbsolute(relation));
}

export function createLocalStore(root) {
  const rootPath = resolve(root);
  return {
    root: rootPath,
    exists(path) {
      return existsSync(resolveAllowedPath(rootPath, path));
    },
    read(path) {
      return readFileSync(resolveAllowedPath(rootPath, path), 'utf8');
    },
    write(path, text) {
      const target = resolveAllowedPath(rootPath, path);
      mkdirSync(dirname(target), { recursive: true });
      writeFileSync(target, String(text ?? ''), 'utf8');
    },
    mkdir(path) {
      const target = resolveAllowedPath(rootPath, `${path}/manifest.json`);
      mkdirSync(dirname(target), { recursive: true });
    },
    remove(path) {
      const target = resolveAllowedPath(rootPath, path);
      if (existsSync(target)) rmSync(target, { force: true });
    },
  };
}

export function readConfigSnapshot(root) {
  const rootPath = resolve(root);
  const rawFiles = {};
  const fileExists = {};
  for (const [platform, filename] of Object.entries({
    squirrel: 'squirrel.custom.yaml',
    weasel: 'weasel.custom.yaml',
    default: 'default.custom.yaml',
  })) {
    const filePath = resolveAllowedPath(rootPath, filename);
    const exists = existsSync(filePath);
    fileExists[platform] = exists;
    rawFiles[platform] = exists ? readFileSync(filePath, 'utf8') : '';
  }
  return {
    folderName: basename(rootPath) || rootPath,
    rawFiles,
    fileExists,
    hasAnySchemaFile: hasSchemaFile(rootPath),
  };
}

export function listBackups(root) {
  const rootPath = resolve(root);
  const backupRoot = resolveAllowedPath(rootPath, `${BACKUP_ROOT}/manifest.json`);
  const backupRootDir = dirname(backupRoot);
  if (!existsSync(backupRootDir)) return [];

  return readdirSync(backupRootDir, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => {
      const backupDir = join(backupRootDir, entry.name);
      const availableFiles = readdirSync(backupDir, { withFileTypes: true })
        .filter((file) => file.isFile())
        .map((file) => file.name)
        .sort((a, b) => a.localeCompare(b));
      const manifestPath = join(backupDir, 'manifest.json');
      const manifest = readManifest(manifestPath);
      return { name: entry.name, manifest, availableFiles };
    })
    .sort((a, b) => b.name.localeCompare(a.name));
}

function readManifest(manifestPath) {
  if (!existsSync(manifestPath)) return null;
  try {
    return JSON.parse(readFileSync(manifestPath, 'utf8'));
  } catch {
    return null;
  }
}

export function createApp({ root, token, staticRoot = STATIC_ROOT }) {
  const rootPath = resolve(root);
  const requiredToken = String(token || '');
  return createServer(async (request, response) => {
    try {
      const url = new URL(request.url || '/', 'http://127.0.0.1');
      if (url.pathname.startsWith('/api/')) {
        if (request.headers['x-rime-editor-token'] !== requiredToken) {
          sendJson(response, 403, { error: '访问令牌无效。' });
          return;
        }
        await handleApi(request, response, url, rootPath);
        return;
      }
      serveStatic(response, staticRoot, url.pathname);
    } catch (error) {
      sendJson(response, 400, { error: error.message || '请求失败。' });
    }
  });
}

export async function startLocalServer({ root = resolve(STATIC_ROOT, '..'), port = 0, openBrowser = true } = {}) {
  const token = randomBytes(18).toString('hex');
  const app = createApp({ root, token });
  await new Promise((resolveListen) => app.listen(port, '127.0.0.1', resolveListen));
  const address = app.address();
  const url = `http://127.0.0.1:${address.port}/?token=${token}`;
  if (openBrowser) openUrl(url);
  return { app, url, root: resolve(root), port: address.port };
}

async function handleApi(request, response, url, root) {
  if (request.method === 'GET' && url.pathname === '/api/config') {
    sendJson(response, 200, readConfigSnapshot(root));
    return;
  }
  if (request.method === 'GET' && url.pathname === '/api/backups') {
    sendJson(response, 200, { backups: listBackups(root) });
    return;
  }
  if (request.method === 'GET' && url.pathname === '/api/file') {
    const path = url.searchParams.get('path');
    const target = resolveAllowedPath(root, path);
    if (!existsSync(target)) {
      sendJson(response, 404, { error: '文件不存在。' });
      return;
    }
    sendJson(response, 200, { exists: true, text: readFileSync(target, 'utf8') });
    return;
  }
  if (request.method === 'PUT' && url.pathname === '/api/file') {
    const body = await readJsonBody(request);
    const target = resolveAllowedPath(root, body.path);
    mkdirSync(dirname(target), { recursive: true });
    writeFileSync(target, String(body.text ?? ''), 'utf8');
    sendJson(response, 200, { ok: true });
    return;
  }
  if (request.method === 'DELETE' && url.pathname === '/api/file') {
    const path = url.searchParams.get('path');
    const target = resolveAllowedPath(root, path);
    if (existsSync(target)) rmSync(target, { force: true });
    sendJson(response, 200, { ok: true });
    return;
  }
  if (request.method === 'POST' && url.pathname === '/api/mkdir') {
    const body = await readJsonBody(request);
    const target = resolveAllowedPath(root, `${body.path}/manifest.json`);
    mkdirSync(dirname(target), { recursive: true });
    sendJson(response, 200, { ok: true });
    return;
  }
  sendJson(response, 404, { error: '接口不存在。' });
}

function serveStatic(response, staticRoot, pathname) {
  const requested = decodeURIComponent(pathname === '/' ? '/index.html' : pathname);
  const relativePath = requested.replace(/^\/+/, '');
  const target = resolve(staticRoot, relativePath);
  const relation = relative(staticRoot, target);
  if (relation === '..' || relation.startsWith(`..${sep}`) || !existsSync(target) || !statSync(target).isFile()) {
    response.writeHead(404, { 'content-type': 'text/plain; charset=utf-8' });
    response.end('Not found');
    return;
  }
  response.writeHead(200, {
    'content-type': MIME_TYPES[extname(target)] || 'application/octet-stream',
    'cache-control': 'no-store',
  });
  createReadStream(target).pipe(response);
}

async function readJsonBody(request) {
  const chunks = [];
  let size = 0;
  for await (const chunk of request) {
    size += chunk.length;
    if (size > 2 * 1024 * 1024) throw new Error('请求内容过大。');
    chunks.push(chunk);
  }
  if (!chunks.length) return {};
  return JSON.parse(Buffer.concat(chunks).toString('utf8'));
}

function sendJson(response, status, value) {
  response.writeHead(status, {
    'content-type': 'application/json; charset=utf-8',
    'cache-control': 'no-store',
  });
  response.end(JSON.stringify(value));
}

function hasSchemaFile(root) {
  return readdirSync(root, { withFileTypes: true }).some((entry) => entry.isFile() && entry.name.endsWith('.schema.yaml'));
}

function openUrl(url) {
  const command = process.platform === 'darwin'
    ? 'open'
    : process.platform === 'win32'
      ? 'cmd'
      : 'xdg-open';
  const args = process.platform === 'win32' ? ['/c', 'start', '', url] : [url];
  const child = spawn(command, args, { stdio: 'ignore', detached: true });
  child.unref();
}

if (process.argv[1] && resolve(process.argv[1]) === MODULE_PATH) {
  const rootArgIndex = process.argv.indexOf('--root');
  const root = rootArgIndex >= 0 ? process.argv[rootArgIndex + 1] : resolve(STATIC_ROOT, '..');
  const portArgIndex = process.argv.indexOf('--port');
  const port = portArgIndex >= 0 ? Number(process.argv[portArgIndex + 1]) : 0;
  startLocalServer({ root, port, openBrowser: !process.argv.includes('--no-open') })
    .then(({ url, root: startedRoot }) => {
      console.log(`Rime 皮肤编辑器已启动：${url}`);
      console.log(`配置目录：${startedRoot}`);
      console.log('关闭这个窗口即可停止本地服务。');
    })
    .catch((error) => {
      console.error(`启动失败：${error.message}`);
      process.exit(1);
    });
}
